worlds/kh1/Rules.py

```diff
@@ -0,0 +1,1962 @@
+                state.has("Progressive Fire", player)
+                and state.has("Crystal Trident", player)
+            )
+        multiworld.get_location("Atlantica Triton's Palace White Trinity Chest"                            , player).access_rule = lambda state: (
+                state.has("White Trinity", player)
+            )
+        multiworld.get_location("Atlantica Defeat Ursula I Mermaid Kick Event"                             , player).access_rule = lambda state: (
+                has_offensive_magic(state, player)
+                and state.has("Crystal Trident", player)
+            )
+        multiworld.get_location("Atlantica Defeat Ursula II Thunder Event"                                 , player).access_rule = lambda state: (
+                state.has("Mermaid Kick", player)
+                and has_offensive_magic(state, player)
+                and state.has("Crystal Trident", player)
+            )
+        multiworld.get_location("Atlantica Seal Keyhole Crabclaw Event"                                    , player).access_rule = lambda state: (
+                state.has("Mermaid Kick", player)
+                and has_offensive_magic(state, player)
+                and state.has("Crystal Trident", player)
+            )
+        multiworld.get_location("Atlantica Undersea Gorge Blizzard Clam"                                   , player).access_rule = lambda state: (
+                state.has("Progressive Blizzard", player)
+            )
+        multiworld.get_location("Atlantica Undersea Valley Fire Clam"                                      , player).access_rule = lambda state: (
+                state.has("Progressive Fire", player)
+            )
+        multiworld.get_location("Atlantica Triton's Palace Thunder Clam"                                   , player).access_rule = lambda state: (
+                state.has("Progressive Thunder", player)
+            )
+        multiworld.get_location("Atlantica Cavern Nook Clam"                                               , player).access_rule = lambda state: (
+                state.has("Crystal Trident", player)
+            )
+        multiworld.get_location("Atlantica Defeat Ursula II Ansem's Report 3"                              , player).access_rule = lambda state: (
+                state.has("Mermaid Kick", player)
+                and has_offensive_magic(state, player)
+                and state.has("Crystal Trident", player)
+            )
+    if options.cups:
+        multiworld.get_location("Olympus Coliseum Defeat Hades Ansem's Report 8"                           , player).access_rule = lambda state: (
+                state.has("Phil Cup", player)
+                and state.has("Pegasus Cup", player)
+                and state.has("Hercules Cup", player)
+                and has_x_worlds(state, player, 7, options.keyblades_unlock_chests)
+                and has_defensive_tools(state, player)
+                and state.has("Entry Pass", player)
```

Comment:
These lines make it look like you were not super aware of the existence of `state.has_all`.
It'd certainly make your code a lot shorter to use it:


```
state.has_all("Phil Cup", "Pegasus Cup", "Hercules Cup", "Entry Pass")
and has_x_worlds(...)
and has_defensive_tools(...)
```

There are many other places where you could do this.

*low priority*